content_scripts/mode_insert.coffee

```diff
@@ -53,6 +56,21 @@ class InsertMode extends Mode
       "focus": (event) => @alwaysContinueBubbling =>
         if @insertModeLock != event.target and DomUtils.isFocusable event.target
           @activateOnElement event.target
+        else if event.target.shadowRoot
+          # A focusable element inside the shadow DOM might have been selected. If so, we can catch the focus
+          # event inside the shadow DOM. This fixes #853.
+          shadowRoot = event.target.shadowRoot
+          eventListeners = {}
+          for type in ["focus", "blur"]
+            eventListeners[type] = do (type) ->
+              (event) ->
+                handlerStack.bubbleEvent type, event
+                if type == "blur"
```

Comment:
@mrmr1993.  What if there are two or more inputs inside the shadow DOM, and one is first focused, then another (and perhaps then another too)?  Wouldn't the `blur` events cause these listeners to be removed before the shadow DOM itself has lost the focus?  Might b40647c48ae7d8016aa77c418eed42c042000f46 be better?